### PR TITLE
General improvements

### PR DESCRIPTION
**Pull request by jecs**
* Using `Diagonal{SVector{N}}` instead of `SMatrix{N,N}` where applicable.
* Converted some `exp(1i*x::Real)` expressions to `cis(x)`.

### CODE REVIEW

**Comment by jecs** [comment]
New `test` output:
```
Time to pre-calculate the interpolation:  1.206 ms (273 allocations: 30.06 KiB)
Time to evaluate the original     super-Lorenzian Green's function:  2.083 μs (2 allocations: 32 bytes)
Time to evaluate the interpolated super-Lorenzian Green's function:  237.309 ns (2 allocations: 32 bytes)
Test Summary:        | Pass  Total   Time
Greens_Interpolation |  103    103  15.3s
Test Summary:             | Pass  Total   Time
apply_hamiltonian_gbloch! |   50     50  33.3s
Test Summary:                              | Pass  Total   Time
apply_hamiltonian_gbloch! shaped RF pulses |    9      9  14.9s
Test Summary:                                | Pass  Total   Time
apply_hamiltonian_gbloch! ω₀-sweep RF pulses |    8      8  11.4s
Test Summary:                     | Pass  Total  Time
apply_hamiltonian_freeprecession! |   50     50  0.6s
Test Summary:                             | Pass  Total   Time
apply_hamiltonian_graham_superlorentzian! |   10     10  33.8s
Test Summary:          | Pass  Total     Time
Graham_saturation_rate |    2      2  1m24.8s
Test Summary:           | Pass  Total   Time
apply_hamiltonian_sled! |   10     10  19.3s
Time to pre-compute saturation:  26.763369 seconds (673.64 M allocations: 63.400 GiB, 14.15% gc time, 75.77% compilation time)
Time to solve the full gene. Bloch IDE for 100us π-pulse:  5.486 ms (30775 allocations: 12.07 MiB)
Time to solve the linear approximation for 100us π-pulse:  1.454 μs (6 allocations: 720 bytes)
Test Summary:        | Pass  Total   Time
Linear_Approximation |    9      9  43.9s
Test Summary:                       | Pass  Total   Time
apply_hamiltonian_gbloch! gradients |   35     35  20.0s
Test Summary:                                 | Pass  Total   Time
apply_hamiltonian_gbloch! shaped-RF gradients |   34     34  11.4s
Test Summary:                                   | Pass  Total   Time
apply_hamiltonian_gbloch! ω₀-sweep RF gradients |   36     36  19.3s
Test Summary:                               | Pass  Total  Time
apply_hamiltonian_freeprecession! gradients |   36     36  2.7s
Test Summary:           | Pass  Total     Time
Linear_Approx gradients |   12     12  1m52.0s
Test Summary:                                       | Pass  Total  Time
apply_hamiltonian_graham_superlorentzian! gradients |   34     34  3.6s
Test Summary:  | Pass  Total   Time
inversionpulse |   60     60  23.6s
Test Summary: | Pass  Total   Time
cf. Solvers   |    4      4  38.5s
Test Summary:     | Pass  Total   Time
Solvers_Gradients |    9      9  28.6s
Test Summary:     | Pass  Total   Time
MultipleInversion |    6      6  22.2s
Test Summary: | Pass  Total     Time
OCT Gradients |    7      7  1m31.1s
Test Summary: | Pass  Total   Time
NLLS          |   27     27  52.6s
     Testing MRIgeneralizedBloch tests passed 
 ```

Old `test` output:
```
Time to pre-calculate the interpolation:  1.288 ms (273 allocations: 30.06 KiB)
Time to evaluate the original     super-Lorenzian Green's function:  2.074 μs (2 allocations: 32 bytes)
Time to evaluate the interpolated super-Lorenzian Green's function:  237.827 ns (2 allocations: 32 bytes)
Test Summary:        | Pass  Total   Time
Greens_Interpolation |  103    103  15.6s
Test Summary:             | Pass  Total   Time
apply_hamiltonian_gbloch! |   50     50  35.4s
Test Summary:                              | Pass  Total   Time
apply_hamiltonian_gbloch! shaped RF pulses |    9      9  15.8s
Test Summary:                                | Pass  Total   Time
apply_hamiltonian_gbloch! ω₀-sweep RF pulses |    8      8  12.3s
Test Summary:                     | Pass  Total  Time
apply_hamiltonian_freeprecession! |   50     50  0.7s
Test Summary:                             | Pass  Total   Time
apply_hamiltonian_graham_superlorentzian! |   10     10  35.8s
Test Summary:          | Pass  Total     Time
Graham_saturation_rate |    2      2  1m27.6s
Test Summary:           | Pass  Total   Time
apply_hamiltonian_sled! |   10     10  19.5s
Time to pre-compute saturation:  27.290919 seconds (673.64 M allocations: 63.400 GiB, 13.49% gc time, 71.50% compilation time)
Time to solve the full gene. Bloch IDE for 100us π-pulse:  5.513 ms (30775 allocations: 12.07 MiB)
Time to solve the linear approximation for 100us π-pulse:  1.488 μs (6 allocations: 720 bytes)
Test Summary:        | Pass  Total   Time
Linear_Approximation |    9      9  44.3s
Test Summary:                       | Pass  Total   Time
apply_hamiltonian_gbloch! gradients |   35     35  20.2s
Test Summary:                                 | Pass  Total   Time
apply_hamiltonian_gbloch! shaped-RF gradients |   34     34  11.8s
Test Summary:                                   | Pass  Total   Time
apply_hamiltonian_gbloch! ω₀-sweep RF gradients |   36     36  19.7s
Test Summary:                               | Pass  Total  Time
apply_hamiltonian_freeprecession! gradients |   36     36  2.7s
Test Summary:           | Pass  Total     Time
Linear_Approx gradients |   12     12  1m51.5s
Test Summary:                                       | Pass  Total  Time
apply_hamiltonian_graham_superlorentzian! gradients |   34     34  3.8s
Test Summary:  | Pass  Total   Time
inversionpulse |   60     60  24.3s
Test Summary: | Pass  Total   Time
cf. Solvers   |    4      4  39.3s
Test Summary:     | Pass  Total   Time
Solvers_Gradients |    9      9  30.0s
Test Summary:     | Pass  Total   Time
MultipleInversion |    6      6  22.9s
Test Summary: | Pass  Total     Time
OCT Gradients |    7      7  1m35.1s
Test Summary: | Pass  Total   Time
NLLS          |   27     27  49.5s
     Testing MRIgeneralizedBloch tests passed 
```

So about ~10% faster here and there.

File: src/DiffEq_Hamiltonians.jl [hunk 1]
```diff
@@ -427,7 +427,7 @@ end
 function add_partial_derivative!(∂m∂t, m, mfun, p::Tuple{Real,Real,Any,Any,Any,Any,Any,Any,Real,Real,Any}, t, grad_type::grad_B1)
     ω1, B1, ω0, m0s, R1f, R2f, Rx, R1s, T2s, TRF, dG_o_dT2s_x_T2s = p
 
-    f_PSD(τ) = quadgk(ct -> 1 / abs(1 - 3 * ct^2) * (4 / τ / abs(1 - 3 * ct^2) * (exp(- τ^2 / 8 * (1 - 3 * ct^2)^2) - 1) + sqrt(2π) * erf(τ / 2 / sqrt(2) * abs(1 - 3 * ct^2))), 0, 1, order=7)[1]
+    f_PSD(τ) = quadgk(ct -> 1 / abs(1 - 3 * ct^2) * (4 / τ / abs(1 - 3 * ct^2) * (exp(- τ^2 / 8 * (1 - 3 * ct^2)^2) - 1) + √(2π) * erf(τ / 2√2 * abs(1 - 3 * ct^2))), 0, 1, order=7)[1]
```
Comment (JakobAsslaender) [inline src/DiffEq_Hamiltonians.jl:430]: This line seems to cause an error in Julia 1.6 – no idea why
```Julia
ERROR: LoadError: LoadError: syntax: missing comma or ) in argument list
Stacktrace:
  [1] top-level scope
    @ ~/work/MRIgeneralizedBloch.jl/MRIgeneralizedBloch.jl/src/DiffEq_Hamiltonians.jl:430
    ```

File: src/MatrixExp_Hamiltonians.jl [hunk 2]
```diff
@@ -318,84 +318,52 @@ function propagator_linear_inversion_pulse(ω1, T, B1, R2s, _, dR2sdB1, grad_typ
 end
 
 function z_rotation_propagator(rfphase_increment, _)
-    u_rot = @SMatrix [cos(rfphase_increment) -sin(rfphase_increment) 0 0 0 0;
-                      sin(rfphase_increment)  cos(rfphase_increment) 0 0 0 0;
-                                     0                0  1 0 0 0;
-                                     0                0  0 1 0 0;
-                                     0                0  0 0 1 0;
-                                     0                0  0 0 0 1]
+	(sϕ,cϕ) = sincos(rfphase_increment)
```
Comment (JakobAsslaender) [inline src/MatrixExp_Hamiltonians.jl:321]: Do you mind removing the brackets and adding a whitespace `sϕ, cϕ = sincos(rfphase_increment)`, and aligning the line with other code?

File: src/MatrixExp_Hamiltonians.jl [hunk 3]
```diff
@@ -318,84 +318,52 @@ function propagator_linear_inversion_pulse(ω1, T, B1, R2s, _, dR2sdB1, grad_typ
...
-                      0                 0                0 1 0 0                 0                0 0 0 0;
-                      0                 0                0 0 1 0                 0                0 0 0 0;
-                      0                 0                0 0 0 cos(rfphase_increment) -sin(rfphase_increment) 0 0 0 0;
-                      0                 0                0 0 0 sin(rfphase_increment)  cos(rfphase_increment) 0 0 0 0;
-                      0                 0                0 0 0 0                 0                1 0 0 0;
-                      0                 0                0 0 0 0                 0                0 1 0 0;
-                      0                 0                0 0 0 0                 0                0 0 1 0;
-                      0                 0                0 0 0 0                 0                0 0 0 1]
+function z_rotation_propagator(rfphase_increment, ::grad_param)
+	(sϕ,cϕ) = sincos(rfphase_increment)
```
Comment (JakobAsslaender) [inline src/MatrixExp_Hamiltonians.jl:332]: Do you mind removing the brackets and adding a whitespace `sϕ, cϕ = sincos(rfphase_increment)`, and aligning the line with other code?

**Review by JakobAsslaender — CHANGES_REQUESTED** [review]
Many thanks! These are nice edits–I learned a bunch from them! One change resulted in the failure of the Julia v1.6 CI tests. Do you mind looking into this? Plus, I made some cosmetic suggestions – I am picky ;)

File: src/DiffEq_Hamiltonians.jl (hunk 1 above)
Comment (jecs) [reply src/DiffEq_Hamiltonians.jl:430]: Of course! Oh, the square root operator `√()` seems to not be valid Julia 1.6. I can revert those changes.

File: src/MatrixExp_Hamiltonians.jl (hunk 2 above)
Comment (jecs) [reply src/MatrixExp_Hamiltonians.jl:321]: Yes, of course, I do not mind. :-) I am a hard tab aficionado and forgot to enable spaces.

**Comment by jecs** [comment]
@JakobAsslaender Should be good now.

`grep -Rnw . -e "√" ` returns 0 lines.

`grep -Rnw . -e "\t"` returns 0 lines.

**Review by JakobAsslaender — APPROVED** [review]

**Pull request merged**